server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java

```diff
@@ -0,0 +1,309 @@
+
+        return (previousUpdateState == null || previousUpdateState.equals(fileUpdateState) == false);
+    }
+
+    @Override
+    protected void doStart() {
+        // We start the file watcher when we know we are master from a cluster state change notification.
+        // We need this additional flag, since cluster state can change after we've shutdown the service
+        // causing the watcher to start again.
+        this.active = true;
+    }
+
+    @Override
+    protected void doStop() {
+        this.active = false;
+        logger.debug("Stopping file settings service");
+        stopWatcher();
+    }
+
+    @Override
+    protected void doClose() {}
+
+    private boolean currentNodeMaster(ClusterState clusterState) {
+        return clusterState.nodes().getMasterNodeId().equals(clusterState.nodes().getLocalNodeId());
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        ClusterState clusterState = event.state();
+        setWatching(currentNodeMaster(clusterState), initialState);
+        initialState = false;
+    }
+
+    private void setWatching(boolean watching, boolean initialState) {
+        if (watching) {
+            startWatcher(initialState);
+        } else {
+            stopWatcher();
+        }
+    }
+
+    // package private for testing
+    boolean watching() {
+        return this.watchService != null;
+    }
+
+    synchronized void startWatcher(boolean onStartup) {
+        if (watching() || active == false) {
+            // already watching or inactive, nothing to do
+            return;
+        }
+
+        logger.info("starting file settings watcher ...");
+
+        Path settingsDir = operatorSettingsDir();
+
+        /**
+         * We essentially watch for two things:
+         *  - the creation of the operator directory (if it doesn't exist), symlink changes to the operator directory
+         *  - any changes to files inside the operator directory if it exists, filtering for settings.json
+         */
+        try {
+            this.watchService = PathUtils.getDefaultFileSystem().newWatchService();
+            if (Files.exists(settingsDir)) {
+                Path settingsFilePath = operatorSettingsFile();
+                if (Files.exists(settingsFilePath)) {
+                    logger.info("found initial operator settings file [{}], applying...", settingsFilePath);
+                    // we make a distinction here for startup, so that if we had operator settings before the node started
+                    // we would fail startup.
+                    processFileSettings(settingsFilePath, onStartup);
+                }
+                enableSettingsWatcher(settingsDir);
+            } else {
+                logger.info("operator settings directory [{}] not found, will watch for its creation...", settingsDir);
+                enableSettingsWatcher(environment.configFile());
```

Comment:
If stashing the expected operator path as a member instead of environment, then this can be `operatorSettingsDir.getParent()`